### PR TITLE
7 implement data preprocessing

### LINKED ISSUE #7: Implement Data Preprocessing
What it says on the tin: any data preprocessing steps required prior to model training, up to and including tokenisation. See @jack89roberts' comment on #4 for how this was done with the TOFU library.

Tokenisation in the HF transformers' library should just require the correct string being passed to the `AutoTokenizer.from_pretrained` method, see [documentation here](https://huggingface.co/docs/transformers/v4.39.3/en/model_doc/auto#transformers.AutoTokenizer.from_pretrained)

### PR DESCRIPTION
**Pull request by J-Dymond**
Initial implementation of the data_modules code from the tofu repo. Currently I test them using an identity function for the tokenizer and in some cases the formatter. There are three, one for returning QA pairs, one for finetuning which returns a formatted Question--Answer pair for autoregression, and one for a forgetting script which returns a formatted retain example and a formatted forget sample. In all cases the forget answer can be replaced with an "idk" type phrase, which is randomly sampled from a .jsonl file.

### CODE REVIEW

**Review by jack89roberts — COMMENTED** [review]
~~Left a few suggestions to talk through but looking good~~ oops wrong PR

File: src/arcsf/data/data_module.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+def QAformatter_basic(QA: tuple[str]) -> str:
+    """
+    Basic QA formatter which accepts a tuple outputs:
+
+    "Question: [input question]\nAnswer: [input answer]"
+
+    args:
+        - QA: Tuple of question answer pair
+    returns:
+        - full_text: formatted question--answer pair
```
Comment (jack89roberts) [inline src/arcsf/data/data_module.py]: Just for convention/consistency, if writing full docstrings the Googledoc style is:

```suggestion
    Basic QA formatter which accepts a tuple and outputs:
    "Question: [input question]\nAnswer: [input answer]"

    Args:
        qa: Tuple of question answer pair
    
    Returns:
        Formatted question--answer pair
```

(Capitalise `Args` and `Returns`, no `-` before args/returns, whitespace between Args and Returns blocks, returns doesn't need a variable name).

File: src/arcsf/data/data_module.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    synonyms by specifying loss_type as "idk".
+    """
+
+    def __init__(
+        self,
+        data,
+        tokenizer,
+        qa_formatter,
+        split,
+        loss_type,
```
Comment (jack89roberts) [inline src/arcsf/data/data_module.py]: types and docstring to explain what these are expected to be would be helpful, e.g. the possible values for `loss_type` and what that changes. Same for the `__init__` of `FinetuneDataset` and `QAForgetDataset`

File: src/arcsf/data/data_module.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+        question = self.data[idx]["question"]
+        answer = self.qa_formatter(self.data[idx]["answer"])
+
+        inp = self.qa_formatter((question, answer))
+
+        return self.tokenizer(inp)
+
+
+class QAForgetDataSet(Dataset):
```
Comment (jack89roberts) [inline src/arcsf/data/data_module.py]: DataSet -> Dataset for consistency (likely needs updating elsewhere too)

```suggestion
class QAForgetDataset(Dataset):
```

File: src/arcsf/data/data_module.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        self.qa_formatter = qa_formatter
+        self.loss_type = loss_type
+        forget_data, retain_data = data
+
+        split_dict = {"retain": retain_data, "forget": forget_data}
+        self.data = split_dict[split]
+
+        if loss_type == "idk":
+            with open("src/arcsf/data/idk.jsonl") as idk_file:
+                self.idk = idk_file.read().splitlines()
```
Comment (jack89roberts) [inline src/arcsf/data/data_module.py]: This path depends on being in the parent `arc-selective-forgetting` directory (i.e. `src/arcsf/...` is a relative path). I usually use `__file__` to get around this (which gives the absolute path to the current file), which there's an example of for the definition of `TEST_DATA_DIR` in `conftest.py`. But the recommended way seems to be to use `importlib.resources` now, see https://stackoverflow.com/a/20885799

File: src/arcsf/data/data_module.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+            self.answer_sampler = self.idk_sampler
+        else:
+            self.answer_sampler = self.forget_sampler
+
+    def answer_sampler():
+        return None
+
+    def idk_sampler(self, _):
+        rand_pos = torch.randint(0, len(self.idk), (1,)).item()
```
Comment (jack89roberts) [inline src/arcsf/data/data_module.py]: To make this reproducible it should be possible to set the initial random seed for this (e.g. by making a random number generator when the class is created and using it here - not sure how that works in `torch` exactly but in `numpy` you can instantiate a rng).

File: src/arcsf/data/data_module.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+        if loss_type == "idk":
+            with open("src/arcsf/data/idk.jsonl") as idk_file:
+                self.idk = idk_file.read().splitlines()
+            self.answer_sampler = self.idk_sampler
+        else:
+            self.answer_sampler = self.forget_sampler
+
+    def idk_sampler(self, _):
+        rand_pos = torch.randint(0, len(self.idk), (1,)).item()
```
Comment (jack89roberts) [inline src/arcsf/data/data_module.py]: See earlier comment re random seeds

File: src/arcsf/data/data_module.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+        if loss_type == "idk":
+            with open("src/arcsf/data/idk.jsonl") as idk_file:
+                self.idk = idk_file.read().splitlines()
+            self.answer_sampler = self.idk_sampler
+        else:
+            self.answer_sampler = self.forget_sampler
+
+    def answer_sampler():
+        return None
```
Comment (jack89roberts) [inline src/arcsf/data/data_module.py]: this was us messing about earlier, if you want to keep it should raise a `NotImplementedError`

File: src/arcsf/data/data_module.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    Basic QA formatter which accepts a tuple outputs:
+
+    "Question: [input question]\nAnswer: [input answer]"
+
+    args:
+        - QA: Tuple of question answer pair
+    returns:
+        - full_text: formatted question--answer pair
+    """
+    Question, Answer = QA
```
Comment (jack89roberts) [inline src/arcsf/data/data_module.py]: Functions and variables/args should be lowercase by convention, e.g.

- `QAformatter_basic` -> `qa_formatter_basic`
- `QA` -> `qa`
- `Question` -> `question`
- `Answer` -> `answer`

File: src/arcsf/data/data_module.py [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+    question_formatting = "Question: " + Question
+    answer_formatting = "\nAnswer: " + Answer
+    full_text = question_formatting + answer_formatting
+    return full_text
```
Comment (jack89roberts) [inline src/arcsf/data/data_module.py]: Could be a lot shorter using an f-string:

```suggestion
    return f"Question: {question}\nAnswer: {answer}"
```

File: src/arcsf/data/data_module.py [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+
+        return self.tokenizer(input), self.tokenizer(target)
+
+
+class FinetuneDataset(Dataset):
+    """
+    Finetune version of the dataset, __getitem__ returns a sample taken either from
+    retain, forget subsets, or a combination of both. Samples are formatted using a
+    question formatter allowing for autoregression. There is an option to output
+    samples using "I don't know" synonyms by specifying loss_type as "idk".
```
Comment (jack89roberts) [inline src/arcsf/data/data_module.py]: `idk` loss doesn't apply for fine-tuning

File: src/arcsf/data/data_module.py [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+            self.answer_sampler = self.forget_sampler
+
+    def answer_sampler():
+        return None
+
+    def idk_sampler(self, _):
+        rand_pos = torch.randint(0, len(self.idk), (1,)).item()
+        return self.idk[rand_pos]
+
+    def forget_sampler(self, forget_row):
```
Comment (jack89roberts) [inline src/arcsf/data/data_module.py]: Is `forget_sampler` the right name for this, will it always be a forget row (e.g. what is split is retrain)? `get_answer` maybe?

File: src/arcsf/data/data_module.py [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+        self.qa_formatter = qa_formatter
+
+        forget_data, retain_data = data
+
+        split_dict = {
+            "retain": retain_data,
+            "forget": forget_data,
+            "all": ConcatDataset([retain_data, forget_data]),
+        }
+        self.data = split_dict[split]
```
Comment (jack89roberts) [inline src/arcsf/data/data_module.py]: `FinetuneDataset` and `EvalQADataset` both only use a single data split (at least at the moment), would it be easier if they were created only with the data they need (so the data argument was a single dataset not a tuple of forget and retain)? That would then also avoid needing this dict/concat step here, which feels a bit odd (start with whole dataset, split it, combine it again).

File: src/arcsf/data/data_module.py [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+        # forget data works as in above examples
+        forget_row = self.forget_data[idx]
+        forget_question = forget_row["question"]
+
+        forget_answer = self.answer_sampler(forget_row)
+
+        retain = self.qa_formatter((retain_question, retain_answer))
+        forget = self.qa_formatter((forget_question, forget_answer))
+
+        return self.tokenizer(retain), self.tokenizer(forget)
```
Comment (jack89roberts) [inline src/arcsf/data/data_module.py:255]: There will likely need to be more logic than only calling the tokenizer here and in the other classes, e.g. see `convert_raw_data_to_model_format` in the TOFU code, but it may be easier to figure out when integrating this with the fine-tunung/forgetting code.

**Review by jack89roberts — CHANGES_REQUESTED** [review]

File: src/arcsf/data/data_module.py (hunk 1 above)
Comment (J-Dymond) [reply src/arcsf/data/data_module.py]: Sorry, I will fix these!

File: src/arcsf/data/data_module.py (hunk 2 above)
Comment (J-Dymond) [reply src/arcsf/data/data_module.py]: Will make these changes on all of the classes. Wasn't sure on the convention for docstrings in classes.

File: src/arcsf/data/data_module.py (hunk 3 above)
Comment (J-Dymond) [reply src/arcsf/data/data_module.py]: Will make this for the three dataset lasses

File: src/arcsf/data/data_module.py (hunk 4 above)
Comment (J-Dymond) [reply src/arcsf/data/data_module.py]: Perfect will investigate this

File: src/arcsf/data/data_module.py (hunk 9 above)
Comment (J-Dymond) [reply src/arcsf/data/data_module.py]: done

File: src/arcsf/data/data_module.py (hunk 10 above)
Comment (J-Dymond) [reply src/arcsf/data/data_module.py]: Will remove this reference

File: src/arcsf/data/data_module.py (hunk 13 above)
Comment (J-Dymond) [reply src/arcsf/data/data_module.py:255]: Yeah I thought so, this was intended as more of a placeholder, until we know what we want it to do. I thought qa_formatter may be able to handle this logic? as in the TOFU code.

File: src/arcsf/data/data_module.py (hunk 5 above)
Comment (J-Dymond) [reply src/arcsf/data/data_module.py]: have done, created a generator for the class to use

File: src/arcsf/data/data_module.py (hunk 6 above)
Comment (J-Dymond) [reply src/arcsf/data/data_module.py]: done

File: src/arcsf/data/data_module.py (hunk 12 above)
Comment (J-Dymond) [reply src/arcsf/data/data_module.py]: Ok, will investigate this

Comment (J-Dymond) [reply src/arcsf/data/data_module.py]: Addressed this now here and in tests, though `load_tofu` outputs a tuple, so its not necessarily obvious which split we are passing to the class using the index. Also these will now operate differently to the `ForgetDataset`. But I suppose this can be addressed down the line.

File: tests/test_data_module.py [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+    """Check that when using an idk loss, that the targets are correct."""
+    data = get_data(
+        "tofu",
+        granularity="question",
+        stratified=False,
+        forget_random=True,
+        forgotten_author_fraction=1 / 3,
+        forgotten_fact_fraction=1 / 3,
+        random_seed=42,
+    )[0]
```
Comment (jack89roberts) [inline tests/test_data_module.py]: Unless I missed something the `get_data` call here can be replaced with the `data` fixture created earlier:

```suggestion
def test_idk_targets(data):
    """Check that when using an idk loss, that the targets are correct."""
```

File: tests/test_data_module.py [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+    idk_set = EvalQADataset(
+        data,
+        _identity,
+        _identity,
+        loss_type="idk",
+    )
+    with open("src/arcsf/data/idk.jsonl") as idk_file:
```
Comment (jack89roberts) [inline tests/test_data_module.py]: path issue in previous comment also applies here, maybe worth having a `get_idk_responses` function in `arcsf/data` somewhere

File: tests/test_data_module.py (hunk 14 above)
Comment (J-Dymond) [reply tests/test_data_module.py]: Yeah, I think you're right. Made this change

File: src/arcsf/data/data_module.py (hunk 4 above)
Comment (J-Dymond) [reply src/arcsf/data/data_module.py]: Thank you for investigating this...

File: tests/test_data_module.py (hunk 15 above)
Comment (J-Dymond) [reply tests/test_data_module.py]: You looked at this for me, thank you

File: src/arcsf/data/data_module.py (hunk 7 above)
Comment (J-Dymond) [reply src/arcsf/data/data_module.py]: Have removed the `get_answer` placeholder now, so the code is cleaner

File: src/arcsf/data/data_module.py [hunk 16]
```diff
@@ -62,12 +81,25 @@ class EvalQADataset(Dataset):
 
     def __init__(
         self,
-        data,
-        tokenizer,
-        qa_formatter,
-        loss_type,
+        data: Dataset,
+        tokenizer: AutoTokenizer,
+        qa_formatter: callable,
```
Comment (jack89roberts) [inline src/arcsf/data/data_module.py]: Instead of `callable` (which is a built-in python function) the typehint should use `Callable` from `collections.abc`: https://docs.python.org/3/library/typing.html#annotating-callable-objects

(i.e. do `from collections.abc import Callable` and then replace all `callable` with `Callable`. You can also type hint the args and return type of the function if you want).

**Review by jack89roberts — APPROVED** [review]

File: src/arcsf/data/data_module.py (hunk 16 above)
Comment (J-Dymond) [reply src/arcsf/data/data_module.py]: Thank you, I wasn't sure how to typehint a function, I've amended this now.
